kolibri/core/content/utils/annotation.py

```diff
@@ -30,109 +34,368 @@
+                if max_val > min_max["max"]:
+                    min_max["max"] = max_val
+                next_lft = rght + 1
+            elif lft != next_lft:
+                or_queries.append(
+                    and_(
+                        ContentNodeTable.c.tree_id == tree,
+                        ContentNodeTable.c.lft >= min_max["min"],
+                        ContentNodeTable.c.rght <= min_max["max"],
+                    )
+                )
+                min_max = {"min": min_val, "max": max_val}
+                next_lft = rght + 1
+        or_queries.append(
+            and_(
+                ContentNodeTable.c.tree_id == tree,
+                ContentNodeTable.c.lft >= min_max["min"],
+                ContentNodeTable.c.rght <= min_max["max"],
+            )
+        )
+
+    if not or_queries:
+        # No constraints that apply in this range, so therefore this query should always
+        # evaluate to False, because nothing can match it.
+        return select([ContentNodeTable.c.id]).where(False == True)  # noqa E712
+
+    # Return a query that ors each of the constraints
+    return select([ContentNodeTable.c.id]).where(or_(*or_queries))
+
 
 def set_leaf_nodes_invisible(channel_id, node_ids=None, exclude_node_ids=None):
+    """
+    Set nodes in a channel as unavailable.
+    With no additional arguments, this will hide an entire channel.
+    With the additional nodes arguments, it will selectively flag nodes
+    as unavailable, based on the passed in ids, setting them as unavailable if
+    they are in node_ids, or descendants of those nodes, but not in
+    exclude_node_ids or descendants of those nodes.
+    """
     bridge = Bridge(app_name=CONTENT_APP_NAME)
 
+    # SQL Alchemy reference to the content node table
     ContentNodeTable = bridge.get_table(ContentNode)
 
     connection = bridge.get_connection()
 
+    # Initial statement for nodes to update
     update_statement = ContentNodeTable.update().where(
         and_(
             ContentNodeTable.c.kind != content_kinds.TOPIC,
             ContentNodeTable.c.channel_id == channel_id,
         )
     )
+    # To chunk the tree, we first find the full extent of the tree - this gives the
+    # highest rght value for this channel.
+    max_rght = connection.execute(
+        select([func.max(ContentNodeTable.c.rght)]).where(
+            ContentNodeTable.c.channel_id == channel_id,
+        )
+    ).scalar()
+
+    # Start a counter for the while loop
+    i = 0
+
+    # Count the total number of constraints
+    constraint_count = len(node_ids or []) + len(exclude_node_ids or [])
+
+    # Aim for a constraint per batch count of about 250 on average
+    # This means that there will be at most 750 parameters from the constraints
+    # and should therefore also limit the overall SQL expression size.
+
+    dynamic_chunksize = max(
+        1, min(CHUNKSIZE, floor(250 * max_rght / (constraint_count or 1)))
+    )
```

Comment:
Can `max_rght` be negative?

If not, I think this is the same as

```python
min(CHUNKSIZE, ceil(250 * max_rght / (constraint_count or 1)))
```

Reply:
It shouldn't be, no, I was just trying to be systematic and perhaps overly cautious.